src/main/java/com/aodocs/endpoints/auth/authenticator/ExtendedAuthenticator.java

```diff
@@ -19,112 +19,87 @@
  */
 @Log
-public abstract class ExtendedAuthenticator implements Authenticator {
+@Singleton
+public class ExtendedAuthenticator implements Authenticator {
 
-    private static final String BASIC_USER_ATTR = "endpoints-authenticators:basicUser";
-    private static final String EXTENDED_USER_ATTR = "endpoints-authenticators:extendedUser";
+    private final Authenticator delegateAuthenticator;
+    private final Authorizer authorizer;
 
-    //use an object to allow future extensions
-    @Value
-    public static class AuthorizationResult {
-        private final boolean authorized;
+    public ExtendedAuthenticator(Authorizer authorizer) {
+        this(new EndpointsAuthenticator(), authorizer);
     }
-
-    private boolean isSingleton;
-
-    public ExtendedAuthenticator() {
-        this.isSingleton = this.getClass().getAnnotation(Singleton.class) != null;
+   
+    public ExtendedAuthenticator(Authenticator delegate, Authorizer authorizer) {
+        this.delegateAuthenticator = delegate;
+        this.authorizer = authorizer;
     }
 
     @Override
-    public User authenticate(final HttpServletRequest request) throws ServiceUnavailableException {
+    public User authenticate(final HttpServletRequest request) throws ServiceException {
         long start = System.currentTimeMillis();
-        Preconditions.checkState(isSingleton, "Extended authenticators must be singletons, as they cache externally loaded data");
 
         //disable client id checking, the whole point of this authenticator is to bypass it
         Attribute attribute = Attribute.from(request);
         attribute.remove(Attribute.ENABLE_CLIENT_ID_WHITELIST);
-
-        final String token = GoogleAuth.getAuthToken(request);
-        if (token == null)
-            return null;
-        final AuthType authType;
-        if (GoogleAuth.isJwt(token)) {
-            authType = AuthType.JWT;
-        } else if (GoogleAuth.isOAuth2Token(token)) {
-            authType = AuthType.OAUTH2;
-        } else {
-            return null;
-        }
-        final User user = getFromRequestOrCompute(request, BASIC_USER_ATTR, () -> performPrimaryAuthentication(request, authType));
+    
+        final User user = delegateAuthenticator.authenticate(request);
         long standardAuthTime = System.currentTimeMillis() - start;
+        
         //the user could be null at this point because the token does
         //not contain the email scope, or any of the required scopes
         if (user == null) {
             return null;
         }
 
-        AuthInfo authInfo = getAuthInfo(authType, request, token);
-        if (authInfo == null) {
-            return null;
-        }
-
+        //Authorization
         try {
-            ExtendedUser extendedUser = getFromRequestOrCompute(request, EXTENDED_USER_ATTR, () -> getExtendedUser(authInfo, user));
+            ExtendedUser extendedUser = getExtendedUser(getAuthInfo(request), user);
```

Comment:
I would keep the ExtendedUser as a request attribute so it can be used in business logic (outside of authenticator)